src/nodes.coffee

```diff
@@ -276,41 +278,36 @@ exports.Base = class Base
   ast: (o, level) ->
     o = Object.assign {}, o
     o.level = level if level?
-    @makeReturn mark: yes if @isStatement(o) and o.level isnt LEVEL_TOP and o.scope?
-
-    fullAst = @astNode o
-    return fullAst unless fullAst?
-    Object.assign fullAst, @astReturns()
-
-  astNode: (o) ->
     # Every abstract syntax tree node object has four categories of properties:
     # - type, stored in the `type` field and a string like `NumberLiteral`.
-    # - location data, stored in the `loc`, `start`, `end` and `range` fields.
+    # - whether this node is implicitly returned, stored as `returns`.
     # - properties specific to this node, like `parsedValue`.
     # - properties that are themselves child nodes, like `body`.
+    # - location data, stored in the `loc`, `start`, `end` and `range` fields.
     # These fields are all intermixed in the Babel spec; `type` and `start` and
     # `parsedValue` are all top level fields in the AST node object. We have
     # separate methods for returning each category, that we merge together here.
-    Object.assign {}, {type: @astType(o)}, @astProperties(o), @astLocationData()
-
-  # By default, a node class has no specific properties.
-  astProperties: -> {}
+    Object.assign {}, {type: @astType(o)}, @astReturns(o), @astProperties(o, level), @astLocationData()
 
   # By default, a node class’s AST `type` is its class name.
   astType: -> @constructor.name
 
+  # Mark AST nodes that correspond to expressions that (implicitly) return.
+  astReturns: (o) ->
+    if @isStatement(o) and o.level isnt LEVEL_TOP and o.scope?
+      # Call `@makeReturn`, rather than simply returning, in case the node
+      # overrides `@makeReturn`.
+      @makeReturn null, yes
+    if @canBeReturned then {returns: yes} else {}
+
+  # By default, a node class has no specific properties.
+  astProperties: -> {}
+
```

Comment:
I’m trying to always keep these in the same order: type, returns, properties, location data. I’m not sure if this is something worth doing something about, but `astReturns`/`makeReturn` needs to be called before `astProperties`, since sometimes in some `astProperties` methods you call `.ast()` which then generates AST nodes for children; but those children need the `canBeReturned` flag already set by that point in order to properly get the `returns: true` included. This is very non-obvious; perhaps we should refactor to remove calls to `.ast()` from `astProperties`?

Reply:
This wouldn't be an issue if you weren't trying to refactor away the parent template method

That's a big part of the point of the pattern I used - it makes sure that `makeReturn()` gets called at the right time

> perhaps we should refactor to remove calls to `.ast()` from `astProperties`?

That's impossible, `astProperties()` has to generate AST for child nodes, which involves calling `.ast()` on those child nodes

Reply:
> That's a big part of the point of the pattern I used - it makes sure that `makeReturn()` gets called at the right time

Then we can put back that explicit call to `makeReturn` from within `ast`; I moved that into the `astReturns` method more for organization than anything else, because they go together. But `makeReturn` can be in `ast` with a comment explaining that it needs to precede the call to `astProperties`, as `astProperties` calls `ast` for child nodes and those nodes need to have `makeReturn` called beforehand. That’s still worth explaining before someone else moves this around and doesn’t realize that the order matters (as I didn’t, for a long time spent debugging).